internal/core/operations/healthcontroller/executor.go

```diff
@@ -369,3 +385,122 @@ func (ex *Executor) canPerformDownscale(ctx context.Context, scheduler *entities
 
 	return can && !waitingCooldown, "ok"
 }
+
+func (ex *Executor) checkRollingUpdate(
+	ctx context.Context,
+	logger *zap.Logger,
+	scheduler *entities.Scheduler,
+	availableRoomsIDs []string,
+) ([]string, bool) {
+	logger.Debug("checking if system is in the middle of a rolling update of scheduler", zap.String("scheduler.Version", scheduler.Spec.Version))
+	var roomsWithPreviousSchedulerVersion []string
+	for _, roomID := range availableRoomsIDs {
+		room, err := ex.roomStorage.GetRoom(ctx, scheduler.Name, roomID)
+		if err == nil && room.Version != scheduler.Spec.Version {
+			roomsWithPreviousSchedulerVersion = append(roomsWithPreviousSchedulerVersion, roomID)
+		}
+	}
+	logger.Debug(
+		"rooms that did not match current scheduler versions",
+		zap.String("scheduler.Version", scheduler.Spec.Version),
+		zap.Int("rooms", len(roomsWithPreviousSchedulerVersion)),
+	)
+	return roomsWithPreviousSchedulerVersion, len(roomsWithPreviousSchedulerVersion) != 0
+}
+
+func (ex *Executor) performRollingUpdate(
+	ctx context.Context,
+	op *operation.Operation,
+	def *Definition,
+	logger *zap.Logger,
+	scheduler *entities.Scheduler,
+	desiredNumberOfRooms int,
+	availableRoomsIDs []string,
+	roomsWithPreviousSchedulerVersion []string,
+) error {
+	logger.Info("performing rolling update", zap.String("scheduler.Version", scheduler.Spec.Version))
+	maxSurgeAmount, err := ex.roomManager.SchedulerMaxSurge(ctx, scheduler)
+	if err != nil {
+		logger.Error("failed to perform rolling update while getting max surge amount of rooms", zap.Error(err))
+		return err
+	}
+	maxRoomsToSurge := desiredNumberOfRooms + maxSurgeAmount - len(availableRoomsIDs)
+	if len(roomsWithPreviousSchedulerVersion) < maxRoomsToSurge {
+		maxRoomsToSurge = len(roomsWithPreviousSchedulerVersion)
+	}
+	if maxRoomsToSurge <= 0 {
+		maxRoomsToSurge = 1
+	}
```

Comment:
can you expand what scenario each equation covers? for example, when can `maxRoomsToSurge` be negative? if desired is in considering downscale?